host/src/udf_query.rs

```diff
@@ -0,0 +1,187 @@
+
+use std::collections::HashMap;
+
+use datafusion::execution::TaskContext;
+use datafusion_common::{DataFusionError, Result as DataFusionResult};
+use datafusion_sql::parser::{DFParserBuilder, Statement};
+use sqlparser::ast::{CreateFunctionBody, Expr, Statement as SqlStatement, Value};
+use sqlparser::dialect::dialect_from_str;
+
+use crate::{WasmComponentPrecompiled, WasmPermissions, WasmScalarUdf};
+
+/// A [ParsedQuery] contains the extracted UDFs and SQL query string
+#[derive(Debug)]
+pub struct ParsedQuery {
+    /// Extracted UDFs from the query
+    pub udfs: Vec<WasmScalarUdf>,
+    /// SQL query string with UDF definitions removed
+    pub sql: String,
+}
+
+/// Handles the registration and invocation of UDF queries in DataFusion with a
+/// pre-compiled WASM component.
+pub struct UdfQueryParser<'a> {
+    /// Pre-compiled WASM component.
+    /// Necessary to create UDFs.
+    components: HashMap<String, &'a WasmComponentPrecompiled>,
+}
+
+impl std::fmt::Debug for UdfQueryParser<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("UdfQueryParser")
+            .field("session_ctx", &"SessionContext { ... }")
+            .field("components", &self.components)
+            .finish()
+    }
+}
+
+impl<'a> UdfQueryParser<'a> {
+    /// Registers the UDF query in DataFusion.
+    pub fn new(
+        components: HashMap<String, &'a WasmComponentPrecompiled>,
+    ) -> Self {
+        Self { components }
+    }
+
+    /// Parses a SQL query that defines & uses Python UDFs into a [ParsedQuery].
+    pub async fn parse(
+        &self,
+        udf_query: &str,
+        permissions: &WasmPermissions,
+        task_ctx: &TaskContext,
+    ) -> DataFusionResult<ParsedQuery> {
+        let (code, sql, lang) = self.parse_inner(udf_query, task_ctx)?;
+
+        let component = self.components.get(&lang).ok_or_else(|| {
+            DataFusionError::Plan(format!(
+                "no WASM component registered for language: {:?}",
+                lang
+            ))
+        })?;
+
+        let udfs = WasmScalarUdf::new(component, permissions, code).await?;
+        Ok(ParsedQuery { udfs, sql })
+    }
+
+    /// Parse the combined query to extract the chosen UDF language, UDF
+    /// definitions, and SQL statements.
+    fn parse_inner(
+        &self,
+        query: &str,
+        task_ctx: &TaskContext,
+    ) -> DataFusionResult<(String, String, String)> {
+        let options = task_ctx.session_config().options();
+
+        let dialect = dialect_from_str(options.sql_parser.dialect.clone()).expect("valid dialect");
+        let recursion_limit = options.sql_parser.recursion_limit;
+
+        let statements = DFParserBuilder::new(query)
+            .with_dialect(dialect.as_ref())
+            .with_recursion_limit(recursion_limit)
+            .build()?
+            .parse_statements()?;
+
+        let mut udf_code = String::new();
```

Comment:
I think you should make this a `HashMap<String, Vec<String>>` (language->vec of code) because:

1. a user may use multiple languages
2. a user may use NO languages/UDFs (see my comment below)
3. the fact that you concat the string w/o an extra newline is a bit presumptuous, not knowing what the string semantics of the guest language are

I would then call `WasmScalarUdf::new` once per code block like:

```rust
let mut udfs = vec![];
for (lang, blocks) in udf_code {
    let component = self.components.get(&lang).unwrap_or_else(...)?;
    for code in blocks {
        udfs.extend(&mut WasmScalarUdf::new(component, permissions, code).await?);
```